sky/server/middleware_utils.py

```diff
@@ -0,0 +1,148 @@
+
+logger = sky_logging.init_logger(__name__)
+
+
+class WebSocketDecision(enum.Enum):
+    ACCEPT = 'accept'
+    UNAUTHORIZED = 'unauthorized'
+    FORBIDDEN = 'forbidden'
+    ERROR = 'error'
+
+
+def websocket_aware(
+        middleware_cls: Type[starlette.middleware.base.BaseHTTPMiddleware]):
+    """Decorator to adapt BaseHTTPMiddleware to handle WebSockets.
+
+    It assembles an HTTP-style request like the HTTP upgrade request during
+    websocket handshake and then delegates it to the real HTTP middleware.
+    """
+
+    class WebSocketAwareMiddleware:
+        """WebSocket-aware middleware wrapper."""
+
+        def __init__(self, app, *args, **kwargs):
+            self.app = app
+            self.middleware = middleware_cls(*args, app=app, **kwargs)
+
+        async def __call__(self, scope, receive, send):
+            scope_type = scope.get('type')
+            if scope_type == 'http':
+                await self.middleware(scope, receive, send)
+            elif scope_type == 'websocket':
+                await self._handle_websocket(scope, receive, send)
+            elif scope_type == 'lifespan':
+                await self.app(scope, receive, send)
+            else:
+                raise ValueError(f'Invalid scope type: {scope_type}')
+
+        async def dispatch(self, request: fastapi.Request, call_next):
+            return await self.middleware.dispatch(request, call_next)
+
+        async def _handle_websocket(self, scope, receive, send):
+            """Handle websocket connection by delegating to HTTP middleware."""
+            decision = await self._run_websocket_dispatch(scope)
+            if decision == WebSocketDecision.ACCEPT:
+                await self.app(scope, receive, send)
+            elif decision == WebSocketDecision.UNAUTHORIZED:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 4401,
+                    'reason': 'Unauthorized',
+                })
+            elif decision == WebSocketDecision.FORBIDDEN:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 4403,
+                    'reason': 'Forbidden',
+                })
+            else:
+                await send({
+                    'type': 'websocket.close',
+                    'code': 1011,
+                    'reason': 'Internal Server Error',
+                })
+
+        async def _run_websocket_dispatch(self, scope) -> WebSocketDecision:
+            http_scope = self._build_http_scope(scope)
+            http_receive = self._http_receive_adapter()
+            request = fastapi.Request(http_scope, receive=http_receive)
+            call_next_called = False
+            stub_response = fastapi.Response(status_code=http.HTTPStatus.OK)
+
+            async def call_next(req):
+                del req
+                nonlocal call_next_called
+                call_next_called = True
+                return stub_response
```

Comment:
This does mean that if a middleware mutates the response (e.g. `RequestIDMiddleware`), that mutation will be discarded. Which could be confusing - we can call it out in the websocket_aware decorator docstring.